### PR TITLE
i18n support

### PR DESCRIPTION
**Pull request by venjiang**
## i18n
twtext international support

### CODE REVIEW

**Review by JonLundy — APPROVED** [review]
Overall this appears to be a good unobtrusive approach to i18n support. LGTM.

**Review by aaadonai — APPROVED** [review]

File: internal/handlers.go [hunk 1]
```diff
@@ -703,6 +703,8 @@ func (s *Server) TimelineHandler() httprouter.Handle {
 
 		ctx := NewContext(s.config, s.db, r)
 
+		log.Infof("conext.lang=%s", ctx.Lang)
+		log.Infof("user.lang=%s", ctx.User.Lang)
```
Comment (prologic) [inline internal/handlers.go:707]: Maybe this is debugging left over? Probably don't need this?

**Review by prologic — APPROVED** [review]

File: internal/handlers.go (hunk 1 above)
Comment (venjiang) [reply internal/handlers.go:707]: Yes, just delete it.

**Comment by prologic** [comment]
Whoops. something went wrong!

```
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | time="2021-04-08T07:49:48Z" level=info msg="Cache version 1"
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | panic: open ./internal/langs/active.en.toml: no such file or directory
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | panic: open ./internal/langs/active.en.toml: no such file or directory
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | panic: open ./internal/langs/active.en.toml: no such file or directory
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | time="2021-04-08T07:49:48Z" level=info msg="merging store ..."
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    |
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    |
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | panic: open ./internal/langs/active.en.toml: no such file or directory
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    |
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | goroutine 1 [running]:
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | goroutine 1 [running]:
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | github.com/nicksnyder/go-i18n/v2/i18n.(*Bundle).MustLoadMessageFile(...)
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | goroutine 1 [running]:
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | github.com/nicksnyder/go-i18n/v2/i18n.(*Bundle).MustLoadMessageFile(...)
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    |
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | github.com/nicksnyder/go-i18n/v2/i18n.(*Bundle).MustLoadMessageFile(...)
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | goroutine 1 [running]:
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | 	/go/pkg/mod/github.com/nicksnyder/go-i18n/v2@v2.1.2/i18n/bundle.go:66
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | 	/go/pkg/mod/github.com/nicksnyder/go-i18n/v2@v2.1.2/i18n/bundle.go:66
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | github.com/nicksnyder/go-i18n/v2/i18n.(*Bundle).MustLoadMessageFile(...)
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewTranslator(0xc000553498)
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewTranslator(0xc000338538)
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | 	/go/pkg/mod/github.com/nicksnyder/go-i18n/v2@v2.1.2/i18n/bundle.go:66
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | 	/go/pkg/mod/github.com/nicksnyder/go-i18n/v2@v2.1.2/i18n/bundle.go:66
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | 	/src/internal/translator.go:18 +0x1da
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | 	/src/internal/translator.go:18 +0x1da
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewTranslator(0xc00051fd30)
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewTranslator(0xc000967d98)
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewServer(0xf088d1, 0xc, 0xc000533e60, 0x23, 0x23, 0x1, 0x2, 0xc00003c180)
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewServer(0xf088d1, 0xc, 0xc00050fe60, 0x23, 0x23, 0x1, 0x2, 0xc000366120)
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | 	/src/internal/server.go:719 +0xdc6
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | 	/src/internal/translator.go:18 +0x1da
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | 	/src/internal/server.go:719 +0xdc6
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | 	/src/internal/translator.go:18 +0x1da
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewServer(0xf088d1, 0xc, 0xc0004ade60, 0x23, 0x23, 0x1, 0x2, 0xc00003c1e0)
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | main.main()
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | main.main()
twtxt_twtxt.1.wvjli5wwcr7p@dm4.mills.io    | 	/src/cmd/twtd/main.go:250 +0x876
twtxt_twtxt.1.nqao4pygi9h4@dm4.mills.io    | 	/src/cmd/twtd/main.go:250 +0x876
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewServer(0xf088d1, 0xc, 0xc000517e60, 0x23, 0x23, 0x1, 0x2, 0xc00003c1e0)
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | 	/src/internal/server.go:719 +0xdc6
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | main.main()
twtxt_twtxt.1.hpu07pag7gl9@dm4.mills.io    | 	/src/cmd/twtd/main.go:250 +0x876
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | 	/src/internal/server.go:719 +0xdc6
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | main.main()
twtxt_twtxt.1.lzel28xb77ny@dm4.mills.io    | 	/src/cmd/twtd/main.go:250 +0x876
```

I'm going to have to roll this back. Can you look at this? I'm sorry I didn't test a production build of this :/ My bad :/

**Comment by prologic** [comment]
Actually rather than roll this back, I'l just rebuild my production instance from a prevbious commit while you fix this 😀

**Comment by prologic** [comment]
@venjiang ☝️

**Comment by venjiang** [comment]
> @venjiang ☝️
copy ./internal/langs/active.*.toml  these files to the distribution directory.

**Comment by prologic** [comment]
Hmm this isn't working:

```
$ git show
commit 0a2e6fdfed1a93b4af208a615016fd7ba24cdba5 (HEAD -> master)
Author: James Mills <prologic@shortcircuit.net.au>
Date:   Thu Apr 8 19:19:24 2021 +1000

    Add missing lang (i18n) files to Docker image builds

diff --git a/Dockerfile b/Dockerfile
index 5a98294..68d700f 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -29,6 +29,9 @@ COPY ./internal/pages/* ./internal/pages/
 # Copy templates
 COPY ./internal/templates/* ./internal/templates/

+# Copy langs (localization / i18n)
+COPY ./internal/langs/* ./internal/langs/
+
 # Copy sources
 COPY *.go ./
 COPY ./internal/*.go ./internal/
```

Am I missing a build step or somethign?

**Comment by venjiang** [comment]
> Hmm this isn't working:
> 
> ```
> $ git show
> commit 0a2e6fdfed1a93b4af208a615016fd7ba24cdba5 (HEAD -> master)
> Author: James Mills <prologic@shortcircuit.net.au>
> Date:   Thu Apr 8 19:19:24 2021 +1000
> 
>     Add missing lang (i18n) files to Docker image builds
> 
> diff --git a/Dockerfile b/Dockerfile
> index 5a98294..68d700f 100644
> --- a/Dockerfile
> +++ b/Dockerfile
> @@ -29,6 +29,9 @@ COPY ./internal/pages/* ./internal/pages/
>  # Copy templates
>  COPY ./internal/templates/* ./internal/templates/
> 
> +# Copy langs (localization / i18n)
> +COPY ./internal/langs/* ./internal/langs/
> +
>  # Copy sources
>  COPY *.go ./
>  COPY ./internal/*.go ./internal/
> ```
> 
> Am I missing a build step or somethign?

You don't need to do anything else. You're right. Any errors?

**Comment by prologic** [comment]
I _think_ I am. You created a `Taskfile.yml` (_which is nice_), but you didn't take into account how much we use `make` for a lot of the b uild and release processes. I'm _trying_ to fix this now, as I don't have time to learn `tasks` just yet...

**Comment by prologic** [comment]
Hmmm maybe not... (_ignore my previous message-)

**Comment by prologic** [comment]
Error is:

```
$ dks logs 623s4sm1p6mm
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | time="2021-04-08T09:19:41Z" level=warning msg="error loading pod settings: open /data/settings.yaml: no such file or directory"
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | time="2021-04-08T09:19:41Z" level=info msg="updating messages cache"
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | time="2021-04-08T09:19:41Z" level=info msg="Cache version 1"
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | time="2021-04-08T09:19:41Z" level=info msg="merging store ..."
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | panic: open ./internal/langs/active.en.toml: no such file or directory
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    |
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | goroutine 1 [running]:
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | github.com/nicksnyder/go-i18n/v2/i18n.(*Bundle).MustLoadMessageFile(...)
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | 	/go/pkg/mod/github.com/nicksnyder/go-i18n/v2@v2.1.2/i18n/bundle.go:66
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewTranslator(0xc00074a2c0)
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | 	/src/internal/translator.go:18 +0x1da
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | github.com/jointwt/twtxt/internal.NewServer(0xf088d1, 0xc, 0xc00041fe60, 0x23, 0x23, 0x1, 0x2, 0xc00003d380)
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | 	/src/internal/server.go:719 +0xdc6
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | main.main()
twtxt_twtxt.1.623s4sm1p6mm@dm4.mills.io    | 	/src/cmd/twtd/main.go:250 +0x876
```

**Comment by venjiang** [comment]
panic: open ./internal/langs/active.en.toml: no such file or directory
Are you sure you copied these files?

**Comment by prologic** [comment]
I have a question.... Are these being loaded from the file system or from inside the binary using `io/fs` and `embed`?

**Comment by prologic** [comment]
Can you try to build a production binary, move it out of the source directory, i.e: `make && make install` and try to repro this? I'm not sure what's going on right now :/

**Comment by prologic** [comment]
I pushed some commits to master...

**Comment by venjiang** [comment]
The Go-i18n package uses the file system to load the language files, which I will improve later.

**Comment by venjiang** [comment]
> Can you try to build a production binary, move it out of the source directory, i.e: `make && make install` and try to repro this? I'm not sure what's going on right now :/

Let me see.

**Comment by prologic** [comment]
> The Go-i18n package uses the file system to load the language files, which I will improve later.

Oh

Surely it uses the new `embed` package and `io/fs` interface by now?

**Comment by venjiang** [comment]
> Can you try to build a production binary, move it out of the source directory, i.e: `make && make install` and try to repro this? I'm not sure what's going on right now :/

Once the language files are copied, there are no other requirements, this one simply reads the files, and it works perfectly.

**Comment by prologic** [comment]
Yeah this is fine but it breaks the "portability" of the binary. It's an extra step that has to be done by the operator.

I'm testing a patch to load the translation files from the embedded `labgs.LocaleFS` right now... Let's see if I'm successful :)

**Comment by venjiang** [comment]
> Yeah this is fine but it breaks the "portability" of the binary. It's an extra step that has to be done by the operator.
> 
> I'm testing a patch to load the translation files from the embedded `labgs.LocaleFS` right now... Let's see if I'm successful :)

I 'm working for embed.FS...

**Comment by prologic** [comment]
I have a patch that's close, but it isn't loading the `zh-cn` translations :/

```#!diff
diff --git a/internal/langs/messages.go b/internal/langs/messages.go
index 01ca8da..9a31836 100644
--- a/internal/langs/messages.go
+++ b/internal/langs/messages.go
@@ -2,7 +2,7 @@ package langs
 
 import "embed"
 
-// go:generate goi18n merge active.*.toml translate.*.toml
+//go:generate goi18n merge active.*.toml translate.*.toml
 
 //go:embed *.toml
 var LocaleFS embed.FS
diff --git a/internal/server.go b/internal/server.go
index e5a7eed..a254d47 100644
--- a/internal/server.go
+++ b/internal/server.go
@@ -716,7 +716,11 @@ func NewServer(bind string, options ...Option) (*Server, error) {
 	}
 
 	// translator
-	translator := NewTranslator()
+	translator, err := NewTranslator()
+	if err != nil {
+		log.WithError(err).Error("error loading translator")
+		return nil, err
+	}
 
 	tmplman, err := NewTemplateManager(config, translator, blogs, cache, archive)
 	if err != nil {
diff --git a/internal/translator.go b/internal/translator.go
index 6e15f12..bd37974 100644
--- a/internal/translator.go
+++ b/internal/translator.go
@@ -1,27 +1,40 @@
 package internal
 
 import (
+	"fmt"
+	"io/fs"
+
 	"github.com/naoina/toml"
 	"github.com/nicksnyder/go-i18n/v2/i18n"
 	"golang.org/x/text/language"
+
+	"github.com/jointwt/twtxt/internal/langs"
 )
 
 type Translator struct {
 	Bundle *i18n.Bundle
 }
 
-func NewTranslator() *Translator {
+func NewTranslator() (*Translator, error) {
 	// lang
 	bundle := i18n.NewBundle(language.English)
 	bundle.RegisterUnmarshalFunc("toml", toml.Unmarshal)
-	// No need to load active.en.toml since we are providing default translations.
-	bundle.MustLoadMessageFile("./internal/langs/active.en.toml")
-	bundle.MustLoadMessageFile("./internal/langs/active.zh-cn.toml")
-	// bundle.LoadMessageFile("./langs/active.zh-tw.toml")
-	return &Translator{
-		Bundle: bundle,
+
+	buf, err := fs.ReadFile(langs.LocaleFS, "active.en.toml")
+	if err != nil {
+		return nil, fmt.Errorf("error loading en locale: %w", err)
 	}
+	bundle.MustParseMessageFileBytes(buf, "active.en.toml")
 
+	buf, err = fs.ReadFile(langs.LocaleFS, "active.zh-cn.toml")
+	if err != nil {
+		return nil, fmt.Errorf("error loading zh-cn locale: %w", err)
+	}
+	bundle.MustParseMessageFileBytes(buf, "active.zh-cn.toml")
+
+	return &Translator{
+		Bundle: bundle,
+	}, nil
 }
 
 // Translate 翻译
```

It loads `en` okay. Any ideas why? Maybe your patch is similar?

**Comment by prologic** [comment]
The way I test this btw is simple:

```#!sh
$ make install
$ cd ~/tmp
$ twtd -R -O -b 0.0.0.0:8000 -l zh-cn -u http://10.0.0.102:8000/ --cookie-secret asdf --magiclink-secret asdf --api-signing-key asdf
```

**Comment by venjiang** [comment]
// translator.go
func NewTranslator() *Translator {
	// lang
	bundle := i18n.NewBundle(language.English)
	bundle.RegisterUnmarshalFunc("toml", toml.Unmarshal)
	MustLoadMessageFromFS(bundle, langs.LangFS, "active.en.toml")
	MustLoadMessageFromFS(bundle, langs.LangFS, "active.zh-cn.toml")
	return &Translator{
		Bundle: bundle,
	}

}
func MustLoadMessageFromFS(b *i18n.Bundle, fsys fs.FS, path string) {
	if _, err := LoadMessageFromFS(b, fsys, path); err != nil {
		panic(err)
	}
}

// LoadMessageFromFileFS is like LoadMessageFile but instead of reading from the
// hosts operating system's file system it reads from the fs file system.
func LoadMessageFromFS(b *i18n.Bundle, fsys fs.FS, path string) (*i18n.MessageFile, error) {
	buf, err := fs.ReadFile(fsys, path)
	if err != nil {
		return nil, err
	}

	return b.ParseMessageFileBytes(buf, path)
}
// message.go
//go:embed active.*.toml
var LangFS embed.FS

**Comment by prologic** [comment]
Hmmm your implementation isn't functionally different to mine, yet running with `twtd -l zh-cn` doesn't work :/ Why?

**Comment by prologic** [comment]
Also I do prefer my patch over yours in terms of "gracefully" handling errors, rather than just plain `panic()`'ing :)

**Comment by prologic** [comment]
Oh wait, am I stupid? Won't my browser pass a `Accepts-Language: en` and override the `-l zh-cn`?

**Comment by prologic** [comment]
Maybe my patch is fine :) Can you test it?

**Comment by venjiang** [comment]
> Also I do prefer my patch over yours in terms of "gracefully" handling errors, rather than just plain `panic()`'ing :)

I agree, I don't like Panic either, but there must be at least one language file, I'm slacking off.

**Comment by prologic** [comment]
Yup confirmed. It works! 

```#!sh
$ curl -q -s -o - -H 'Accept-Language: zh-cn' http://10.0.0.102:8000/register | grep -A 10 '<main.*>'
  <main class="container">

  <article class="grid">
    <div>
      <hgroup>
          <h2>注册</h2>
        <p> 在twtxt.net上创建并注册一个新的社交帐户</p>
      </hgroup>
      <form id="register" action="/register" method="POST">
        <input type="hidden" name="csrf_token" value="t&#43;xAT9rkfiyr/n5EQMNnSOm1kerU&#43;bzYaASFS87EGhSiBmb5bJDAzAMWpKmDBdqVtFTD&#43;HFdokPHxX9HWqv9OQ==">
        <input type="text" name="username" placeholder="用户名" aria-label="Username" autocomplete="nickname" autofocus required>

```

**Comment by prologic** [comment]
Are you happy with my patch? I'll commit and push this and re-deploy and we should ba happy campers!

**Comment by prologic** [comment]
Pushed

**Comment by venjiang** [comment]
> Hmmm your implementation isn't functionally different to mine, yet running with `twtd -l zh-cn` doesn't work :/ Why?



> Are you happy with my patch? I'll commit and push this and re-deploy and we should ba happy campers!



> Pushed

great job.

**Comment by prologic** [comment]
> great job.

@venjiang All thanks to you buddy! :D Now go spin up an instance, invite all your friends and spread the word ;)

**Comment by prologic** [comment]
And thank you for doing this!

**Comment by venjiang** [comment]
I love this project and will continue to make it better. I'm going to spend some time looking at the source code, and I'm going to understand it completely.

**Comment by prologic** [comment]
Hit me up on [twtxt.net](https://twtxt.net) and tag me if you have any q's. We're also in the processing of re-branding as you know.

**Pull request merged**